src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+                                  MacroAssembler *_masm);
+
+static void byte_compare_helper(int size, Label &L_noMatch, Label &L_matchFound, Register needle, Register needleVal,
+                                Register haystack, Register mask, Register foundIndex, Register tmp,
+                                StrIntrinsicNode::ArgEncoding ae, MacroAssembler *_masm);
+
+static void highly_optimized_short_cases(StrIntrinsicNode::ArgEncoding ae, Register haystack, Register haystack_len,
+                                         Register needle, Register needle_len, XMMRegister XMM0, XMMRegister XMM1,
+                                         Register mask, Register tmp, MacroAssembler *_masm);
+
+static void setup_jump_tables(StrIntrinsicNode::ArgEncoding ae, Label &L_error, Label &L_checkRange, Label &L_fixup,
+                              address *jump_table, address *jump_table_1, MacroAssembler *_masm);
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  //                         Start of generator
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+
+void StubGenerator::generate_string_indexof(address *fnptrs) {
+  assert((int) StrIntrinsicNode::LL < 4, "Enum out of range");
+  assert((int) StrIntrinsicNode::UL < 4, "Enum out of range");
+  assert((int) StrIntrinsicNode::UU < 4, "Enum out of range");
+  generate_string_indexof_stubs(fnptrs, StrIntrinsicNode::LL);
+  generate_string_indexof_stubs(fnptrs, StrIntrinsicNode::UL);
+  generate_string_indexof_stubs(fnptrs, StrIntrinsicNode::UU);
+  assert(fnptrs[StrIntrinsicNode::LL] != nullptr, "LL not generated.");
+  assert(fnptrs[StrIntrinsicNode::UL] != nullptr, "UL not generated.");
+  assert(fnptrs[StrIntrinsicNode::UU] != nullptr, "UU not generated.");
+}
+
+void StubGenerator::generate_string_indexof_stubs(address *fnptrs, StrIntrinsicNode::ArgEncoding ae) {
+  StubCodeMark mark(this, "StubRoutines", "stringIndexOf");
+  bool isLL = (ae == StrIntrinsicNode::LL);
+  bool isUL = (ae == StrIntrinsicNode::UL);
+  bool isUU = (ae == StrIntrinsicNode::UU);
+  bool isU = isUL || isUU;  // At least one is UTF-16
+  assert(isLL || isUL || isUU, "Encoding not recognized");
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  //                         AVX2 code
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  if (VM_Version::supports_avx2()) {  // AVX2 version
```

Comment:
Instead of the if check here, it would be better to do an assert here:
assert (VM_Version::supports_avx2(), "Needs AVX2 support");